src/util/strencodings.h

```diff
@@ -71,9 +71,19 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
  * number of hex digits.*/
 bool IsHex(std::string_view str);
 /**
-* Return true if the string is a hex number, optionally prefixed with "0x"
-*/
-bool IsHexNumber(std::string_view str);
+ * @brief Return a sanitized hex string if input is a valid hex number.
+ *        "0x" prefix is removed, and if result_size is specified, the
+ *        result is padded with leading zeroes until result_size is
+ *        reached, or std::nullopt returned if input is longer than
+ *        result_size.
```

Comment:
> I think instead of saying this "sanitizes" the string and pretending like that word means something it would be better to just say this strips 0x prefixes and fixes the length of the string so it can be passed to the base_blob::ParseHex() function.

Yeah, I felt a bit uneasy about the arbitrariness of what the function is doing and how it's named. I can't really come up with a better name, though, but I'm open to suggestions?

An alternative would be to, as @maflcko initially [suggested](https://github.com/bitcoin/bitcoin/pull/30569/#discussion_r1700712603), just have this be a parameter of `base_blob::FromHex()`, but that feels less elegant to me, still. What do you think?

Reply:
> It's hard to imagine this function being used for anything other than as a preprocessor to `FromHex`

The reason I carved it out instead of having it be a `FromHex()` parameter is not for reusability but to not bloat `FromHex()` with an explosion of input validation options (which also leaks into functions wrapping `base_blob::FromHex()`, such as `transaction_identifier::FromHex()`). I don't think there is a single answer to what we'll ever need for e.g. the below options, it can very much vary for each callsite and context:
- do we accept leading and/or trailing whitespaces
- do we require fixed size input
- do we forbid, allow or require a 0x prefix
- is an empty string forbidden or equivalent to 0

I think having an opinionated input validation function (and potentially more in the future, when the need arises) and a pristine `FromHex()` that does not concern itself with input validation is a much more maintainable and testable approach. For that reason, I'd prefer not parameterizing `FromHex()`.

Would an approach like the below (moved into the `uint256.h` header), alleviate your concern?

```
/**
 * @brief Helper function for \ref base_blob::FromHex. Returns a copy of
 *        the input without (optional) "0x" prefix which is padded with
 *        leading zeroes if it is shorter than what T::FromHex expects.
 *        If input contains invalid hex characters, including
 *        whitespace, or if it is longer than what T::FromHex expects, a
 *        std::nullopt is returned.
 *
 *        If a value is returned, it will be accepted by T::FromHex().
 *
 * @param input Hex encoding of a number, optionally prefixed with 0x.
 */
template <typename T>
std::optional<std::string> ValidateAndPadHexNumber(std::string_view input) 
    requires std::is_base_of_v<base_blob<T::size() * 8>, T>;
```

Reply:
I think I would probably add `base_blob::FromUserHex()` function wrapping `base_blob::FromHex()`. `FromHex()` would require strict formatting with fixed width strings and no extraneous characters. But when provided with user input, `FromUserHex()` would allow `0x` prefixes, and not require numbers to be padded to a fixed number of digits. Either way, numbers that are too big to fit in the blob or contained unrecognized characters would be errors.

There is no need for either of these functions to trim whitespace as we already have convenient functions to do that, and decision on whether to allow whitespace depends more on how the string arrived (from a file, or over the command line or RPC) than on what produced it.

I think a `FromUserHex` function would be better than a sidecar `ValidateAndPadHexNumber` function because it would be easier to use, and more discoverable. The more it was used, the more it would provide a consistent user interface that did not arbitrarily reject things some places which are accepted other places.

Reply:
> I think I would probably add base_blob::FromUserHex() function wrapping base_blob::FromHex().

Thanks for the suggestion, I've adopted this approach.